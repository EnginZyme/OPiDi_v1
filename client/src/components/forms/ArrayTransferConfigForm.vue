<template>
  <div class="step_config_form" v-if="formType === 'array_transfer'">
    <p class="step_config_form__title">Array Transfer</p>
    <div class="step_config_form__body">
      <div class="step_config_form__body__content">
        <div>
          <p class="step_config_form__body__content__label--medium">
            Step Name
          </p>
          <InputText
            type="text"
            class="step_config_form__body__content__input--long"
            v-model="arrayTransferStepName"
          />
        </div>
        <div>
          <p class="step_config_form__body__content__label--medium">Pipette</p>
          <Dropdown
            :options="
              cachedProtocolPipettes.filter((item) =>
                item.name.toLowerCase().includes('single')
              )
            "
            optionLabel="name"
            placeholder="Select option"
            class="step_config_form__body__content__dropdown"
            v-model="arrayTransferPipette"
          />
        </div>
        <div>
          <p class="step_config_form__body__content__label--medium">
            Tips Strategy
          </p>
          <Dropdown
            :options="[
              { name: 'Standard', value: 'STANDARD' },
              { name: 'Strict', value: 'STRICT' },
              { name: 'Eco', value: 'ECO' },
              { name: 'Eco - New Tip', value: 'ECO - NEW TIP' },
            ]"
            optionLabel="name"
            v-tooltip="
              'STANDARD: one tip per source well, STRICT: one tip per dispense, ECO: reuse tip'
            "
            placeholder="Select option"
            class="step_config_form__body__content__dropdown"
            v-model="arrayTransferTipsStrategy"
          />
        </div>
        <div>
          <p class="step_config_form__body__content__label--medium">
            Liquid Class
          </p>
          <Dropdown
            :options="
              cachedProtocolLiquidClasses.map((item) => ({
                name: item.name,
                id: item.id,
              }))
            "
            optionLabel="name"
            placeholder="Select option"
            class="step_config_form__body__content__dropdown"
            v-model="arrayTransferLiquidConfig"
          />
        </div>

        <div>
          <p class="step_config_form__body__content__label--medium">
            Source Offset
          </p>
          <div class="step_config_form__body__content__field_group">
            <div>
              <p class="step_config_form__body__content__label--min">Type</p>
              <Dropdown
                :options="[{ name: 'Top' }, { name: 'Bottom' }]"
                optionLabel="name"
                placeholder="Select option"
                class="step_config_form__body__content__dropdown"
                v-model="arrayTransferSourceOffsetType"
              />
            </div>
            <div>
              <p class="step_config_form__body__content__label--min">
                Value (mm)
              </p>
              <InputNumber
                :minFractionDigits="2"
                :maxFractionDigits="5"
                class="step_config_form__body__content__input--short"
                v-model="arrayTransferSourceOffset"
              />
            </div>
          </div>
        </div>
        <div>
          <p class="step_config_form__body__content__label--medium">
            Destination Offset
          </p>
          <div class="step_config_form__body__content__field_group">
            <div>
              <p class="step_config_form__body__content__label--min">Type</p>
              <Dropdown
                :options="[{ name: 'Top' }, { name: 'Bottom' }]"
                optionLabel="name"
                placeholder="Select option"
                class="step_config_form__body__content__dropdown"
                v-model="arrayTransferDestinationOffsetType"
              />
            </div>
            <div>
              <p class="step_config_form__body__content__label--min">
                Value (mm)
              </p>
              <InputNumber
                :minFractionDigits="2"
                :maxFractionDigits="5"
                class="step_config_form__body__content__input--short"
                v-model="arrayTransferDestinationOffset"
              />
            </div>
          </div>
        </div>
        <div>
          <p class="step_config_form__body__content__label--medium">
            Array Map Filename
          </p>
          <div class="step_config_form__body__content__file_group">
            <InputText
              class="step_config_form__body__content__input--stretch"
              v-model="arrayMapFileName"
              readonly
            />
            <Button
              label="Browse"
              class="step_config_form__body__content__file_upload_button"
              @click="requestUploadFile"
              for="uploadmyfile"
            />
            <input
              type="file"
              id="uploadArrayMapFile"
              ref="uploadArrayMapFile"
              @change="parseArrayMapFile"
              accept=".csv"
            />
          </div>
        </div>
        <a
          class="step_config_form__body__content__file_hyperlink"
          href="/assets/array_map_template_simplified.xlsx"
          download
          >Download Template</a
        >
        <Button
          :label="arrayMappingButttonLabel"
          :class="arrayMappingButttonClass"
          class="step_config_form__body__content__colour_button"
          @click="shouldDisplayArrayMapModal = true"
          :disabled="
            assertValidDestinationVariableMap(getProtocolStep(stepID).parameters.destination_var_map)
          "
        />
      </div>
    </div>
    <ArrayMapForm
      :shouldDisplayModal="shouldDisplayArrayMapModal"
      :stepID="stepID"
      @formClosed="closeArrayMapModal"
      :pipette="arrayTransferPipette"
    />
  </div>
</template>

<script>
import types from "../../types";
import { mapGetters } from "vuex";
import Papa from "papaparse";
import { findDeep } from "../../utils";
import ArrayMapForm from "./ArrayMapForm.vue";

/**
 * A component to receive user input on the configuration of a chosen array transfer step.
 */
export default {
  components: { ArrayMapForm },
  name: "ArrayTransferConfigForm",
  props: {
    /**
     * Indicates if this component should be displayed or not (true if its value is set to `array_transfer`).
     */
    formType: String,
    /**
     * The ID of the current user clicked step in the protocol editor UI.
     */
    stepID: Number,
  },
  data() {
    return {
      /**
       * Determines if the array map modal child component is displayed or not.
       */
      shouldDisplayArrayMapModal: false,
    };
  },
  methods: {
    /**
     * Parses the input for the array map file for this step.
     */
    parseArrayMapFile() {
      let temp = this;
      let file = this.$refs.uploadArrayMapFile.files[0];
      this.$refs.uploadArrayMapFile.value = null;

      function isValidFile(arr) {
        if (
          arr.length == 0 ||
          arr[0].length < 3 ||
          !arr[0][0].includes("Labware Name") ||
          !arr[0][1].includes("Well Location") ||
          !arr[0][2].includes("Well Volume")
        )
          return false;
        return true;
      }

      Papa.parse(file, {
        skipEmptyLines: true,
        complete: function(results, file) {
          if (!isValidFile(results.data)) {
            temp.$toast.add({
              severity: "error",
              summary: "Notification",
              detail: "Unrecognized input file format",
              life: 1700,
            });
            return;
          }
          temp.$store.dispatch(types.UPDATE_ARRAY_MAP_FILENAME, {
            id: temp.stepID,
            newValue: file.name,
          });
          temp.$store.dispatch(types.UPDATE_ARRAY_MAP_PARSED_FILE, {
            id: temp.stepID,
            newValue: results.data,
          });
        },
      });
    },
    /**
     * Closes the array map modal child component.
     */
    closeArrayMapModal() {
      this.shouldDisplayArrayMapModal = false;
    },
    /**
     * Fetches the mapped protocol step from list of all steps by its ID.
     * 
     * @param {number} id - The ID of the said protocol step.
     */
    getProtocolStep(id) {
      return findDeep(this.cachedProtocolSteps, id);
    },
    /**
     * Simulates a user click on DOM element to initiate a file upload.
     */
    requestUploadFile() {
      this.$refs.uploadArrayMapFile.click();
    },
    /**
     * Asserts that a destination variable map has the accepted form.
     * 
     * @param {object} varMap - The said destination variable map.
     */
    assertValidDestinationVariableMap(varMap) {
      if (this.formType === "array_transfer" && this.stepID >= 0) {
        return Object.keys(varMap).length === 0 || !this.arrayTransferPipette;
      }
      return false;
    },
  },
  computed: {
    ...mapGetters([
      "cachedProtocolSteps",
      "cachedProtocolPipettes",
      "nonTipRackLabwares",
      "cachedProtocolLiquidClasses",
    ]),
    /**
     * Defines the class attribute of the array mapping button.
     */
    arrayMappingButttonClass() {
      if (this.formType === "array_transfer" && this.stepID >= 0) {
        return this.getProtocolStep(this.stepID).parameters.hasMapConflict ||
          !this.getProtocolStep(this.stepID).parameters.mappingComplete
          ? ["p-button-raised", "p-button-secondary"]
          : ["p-button-raised", "p-button-success"];
      }
      return "";
    },
    /**
     * Defines the label attribute of the array mapping button.
     */
    arrayMappingButttonLabel() {
      if (this.formType === "array_transfer" && this.stepID >= 0) {
        return this.getProtocolStep(this.stepID).parameters.hasMapConflict ||
          !this.getProtocolStep(this.stepID).parameters.mappingComplete
          ? "Define Array Mapping"
          : "Update Array Mapping";
      }
      return "";
    },
    /**
     * A read/write object for the array transfer step name.
     */
    arrayTransferStepName: {
      get: function() {
        return this.formType !== "array_transfer" || this.stepID < 0
          ? null
          : this.getProtocolStep(this.stepID).name;
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_TRANSFER_STEP_NAME, {
          id: this.stepID,
          newValue,
        });
      },
    },
    /**
     * A read/write object for the array transfer pipette.
     */
    arrayTransferPipette: {
      get: function() {
        return this.formType !== "array_transfer" || this.stepID < 0
          ? null
          : this.cachedProtocolPipettes.filter(
              (item) =>
                item.name ===
                this.getProtocolStep(this.stepID).parameters.pipette_obj.name
            )[0];
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_TRANSFER_PIPETTE, {
          id: this.stepID,
          newValue,
        });
      },
    },
    /**
     * A read/write object for the array transfer pipette strategy.
     */
    arrayTransferPipetteStrategy: {
      get: function() {
        return this.formType !== "array_transfer" ||
          this.stepID < 0 ||
          !this.getProtocolStep(this.stepID).parameters.pipette_strategy
          ? null
          : {
              value: this.getProtocolStep(this.stepID).parameters.pipette_strategy,
              name: this.getProtocolStep(this.stepID)
                .parameters.pipette_strategy.split("_")
                .map(
                  (string) =>
                    string.charAt(0).toUpperCase() +
                    string.slice(1).toLowerCase()
                )
                .join(" "),
            };
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_TRANSFER_PIPETTE_STRATEGY, {
          id: this.stepID,
          newValue: newValue.value,
        });
      },
    },
    /**
     * A read/write object for the array transfer tips strategy.
     */
    arrayTransferTipsStrategy: {
      get: function() {
        return this.formType !== "array_transfer" ||
          this.stepID < 0 ||
          !this.getProtocolStep(this.stepID).parameters.tipsStrategy
          ? null
          : {
              value: this.getProtocolStep(this.stepID).parameters.tipsStrategy,
              name: this.getProtocolStep(this.stepID)
                .parameters.tipsStrategy.toLowerCase()
                .replace(/\b\w/g, (l) => l.toUpperCase()),
            };
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_TRANSFER_TIPS_STRATEGY, {
          id: this.stepID,
          newValue: newValue.value,
        });
      },
    },
    /**
     * A read/write object for the array transfer liquid configuration.
     */
    arrayTransferLiquidConfig: {
      get: function() {
        return this.formType !== "array_transfer" ||
          this.stepID < 0 ||
          typeof this.getProtocolStep(this.stepID).parameters.liquidClassId !==
            "number"
          ? null
          : {
              id: this.getProtocolStep(this.stepID).parameters.liquidClassId,
              name: this.getProtocolStep(this.stepID).parameters.liquidClass,
            };
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_TRANSFER_LIQUID_CLASS, {
          id: this.stepID,
          value: newValue,
        });
      },
    },
    /**
     * A read/write object for the array transfer source offset type.
     */
    arrayTransferSourceOffsetType: {
      get: function() {
        return this.formType !== "array_transfer" || this.stepID < 0
          ? null
          : { name: this.getProtocolStep(this.stepID).parameters.offset.source_type };
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_TRANSFER_SOURCE_OFFSET_TYPE, {
          id: this.stepID,
          newValue: newValue.name,
        });
      },
    },
    /**
     * A read/write object for the array transfer destination offset type.
     */
    arrayTransferDestinationOffsetType: {
      get: function() {
        return this.formType !== "array_transfer" || this.stepID < 0
          ? null
          : {
              name: this.getProtocolStep(this.stepID).parameters.offset
                .destination_type,
            };
      },
      set: function(newValue) {
        this.$store.dispatch(
          types.UPDATE_ARRAY_TRANSFER_DESTINATION_OFFSET_TYPE,
          {
            id: this.stepID,
            newValue: newValue.name,
          }
        );
      },
    },
    /**
     * A read/write object for the array transfer source offset.
     */
    arrayTransferSourceOffset: {
      get: function() {
        return this.formType !== "array_transfer" || this.stepID < 0
          ? null
          : this.getProtocolStep(this.stepID).parameters.offset.source;
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_TRANSFER_SOURCE_OFFSET, {
          id: this.stepID,
          newValue: parseFloat(newValue),
        });
      },
    },
    /**
     * A read/write object for the array transfer destination offset.
     */
    arrayTransferDestinationOffset: {
      get: function() {
        return this.formType !== "array_transfer" || this.stepID < 0
          ? null
          : this.getProtocolStep(this.stepID).parameters.offset.destination;
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_TRANSFER_DESTINATION_OFFSET, {
          id: this.stepID,
          newValue: parseFloat(newValue),
        });
      },
    },
    /**
     * A read/write object for the array transfer step's array map filename.
     */
    arrayMapFileName: {
      get: function() {
        return this.formType !== "array_transfer" || this.stepID < 0
          ? null
          : this.getProtocolStep(this.stepID).parameters.array_map_filename;
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_ARRAY_MAP_FILENAME, {
          id: this.stepID,
          newValue,
        });
      },
    },
  },
};
</script>

<style scoped>
.step_config_form {
  display: grid;
  grid-template-rows: auto auto;
  padding-top: 20px;
  padding-bottom: 10px;
}
.step_config_form__body {
  border-radius: 5px;
  box-shadow: 0 -1em 1em -1em #bcbcbc inset;
  padding-top: 20px;
  padding-left: 16px;
  display: grid;
  overflow: auto;
  height: 65vh;
  margin-top: 10px;
}
.step_config_form__body__content {
  align-self: center;
  display: grid;
  grid-row-gap: 3vh;
  padding-bottom: 30px;
}
.step_config_form__body__content__file_hyperlink {
  font-size: 9px;
}
.step_config_form__body__content__colour_button {
  width: 91%;
  font-size: 14px;
  font-weight: bold;
}
.step_config_form__body__content__input--short {
  background: #efefff;
  box-shadow: inset 2px 2px 4px rgba(0, 0, 0, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25);
  border-radius: 6px;
  font-size: 12px;
  width: 80%;
  height: 32px;
  border: none;
}
.step_config_form__body__content__input--stretch {
  background: #efefff;
  box-shadow: inset 2px 2px 4px rgba(0, 0, 0, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25);
  border-radius: 6px;
  font-size: 12px;
  height: 32px;
  border: none;
}
.step_config_form__body__content__input--long {
  background: #efefff;
  box-shadow: inset 2px 2px 4px rgba(0, 0, 0, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25);
  border-radius: 6px;
  font-size: 12px;
  width: 91%;
  height: 32px;
  border: none;
}
#uploadArrayMapFile {
  display: none;
}
.step_config_form__body__content__file_upload_button {
  background: #efefff;
  border-radius: 4px;
  border: 0;
  color: grey;
  height: 30px;
  font-size: 14px;
  font-weight: bold;
  filter: drop-shadow(3px 3px 4px rgba(0, 0, 0, 0.25));
}
.step_config_form__body__content__file_group {
  display: grid;
  grid-template-columns: 6fr auto;
  grid-column-gap: 8px;
  width: 91%;
}
.step_config_form__title {
  font-size: 18px;
  margin: 0px;
}
.step_config_form__body__content__field_group {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-column-gap: 40px;
  width: 92%;
}
.step_config_form__body__content__label--medium {
  font-size: 13px;
  margin: 0px;
  margin-bottom: 8px;
}
.step_config_form__body__content__dropdown {
  background: #efefff;
  width: 91%;
  font-size: 9px;
  box-shadow: 2px 2px 4px 1px rgba(0, 0, 0, 0.25),
    -2px -2px 4px 1px rgba(255, 255, 255, 0.25);
  border-radius: 6px;
}
.step_config_form__body__content__label--min {
  font-size: 10px;
  margin: 0px;
  margin-bottom: 8px;
}
</style>
