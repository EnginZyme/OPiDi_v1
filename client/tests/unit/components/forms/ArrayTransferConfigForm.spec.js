import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import '../../matchMedia';
import ArrayTransferConfigForm from "@/components/forms/ArrayTransferConfigForm";

const localVue = createLocalVue();

localVue.use(Vuex);

const sampleFormType = "array_transfer";
const sampleStepID = 8;
const sampleCachedProtocolSteps = [
  {
    type: "array_transfer",
    name: "Array Transfer (8)",
    id: 8,
    substeps: [],
    parameters: {
      pipette: "right_pipette",
      pipette_obj: {
        name: "(Right) Single Channel 300 uL - Gen 1",
        category: "right_pipette",
      },
      pipette_strategy: "",
      array_map_filename: "demo.csv",
      source_var_map: {
        "buffer 1": {
          slot_name: "tuberack_2",
          slot_number: 2,
          wells: ["A3"],
          destinations: {
            destination_plate: {
              volumes: ["500", "0"],
              wells: ["A1", "B1"],
            },
          },
        },
        "buffer 2": {
          slot_name: "tuberack_2",
          slot_number: 2,
          wells: ["A4"],
          destinations: {
            destination_plate: {
              volumes: ["0", "500"],
              wells: [],
            },
          },
        },
        water: {
          slot_name: "reservoir_5",
          slot_number: 5,
          wells: ["A1"],
          destinations: {
            destination_plate: {
              volumes: ["1000", "1000"],
              wells: ["A1", "B1"],
            },
          },
        },
      },
      destination_var_map: {
        destination_plate: { slot_name: "wellplate_1", slot_number: 1 },
      },
      hasMapConflict: false,
      mappingComplete: true,
      array_map: [
        ["tuberack_2", "A3", "wellplate_1", "A1", "500"],
        ["tuberack_2", "A3", "wellplate_1", "B1", "0"],
        ["tuberack_2", "A4", "wellplate_1", "A1", "0"],
        ["tuberack_2", "A4", "wellplate_1", "B1", "500"],
        ["reservoir_5", "A1", "wellplate_1", "A1", "1000"],
        ["reservoir_5", "A1", "wellplate_1", "B1", "1000"],
      ],
      liquidClass: "Default",
      liquidClassId: 0,
      tipsStrategy: "ECO",
      offset: {
        source: 1,
        destination: 1,
        source_type: "Bottom",
        destination_type: "Bottom",
      },
    },
  },
];
const sampleCachedProtocolLiquidClasses = [
  {
    name: "Default",
    id: 0,
    liquid_config_object: {
      touch_tip: false,
      blow_out: true,
      carryover: true,
      air_gap: 0,
      blowout_location: "destination well",
      mix_before: { repetitions: 0, volume: 0 },
      mix_after: { repetitions: 0, volume: 0 },
      aspirate_rate: 1,
      dispense_rate: 1,
      blow_out_rate: 1,
    },
  },
];
const sampleNonTipRackLabwares = [
  {
    name: "wellplate_1",
    labware_name: "test 96 wellplate 2ml 3xT",
    labware_type: "shakerLabware",
    labware_id: "test_96_wellplate_2ml_3xT",
    nickname: "",
    display_name: "test 96 wellplate 2ml 3xT [1]",
    id: 1,
  },
  {
    name: "tuberack_2",
    labware_name: "test tuberack 65 mL tubes",
    labware_type: "tubeRack",
    labware_id: "test_tuberack_65_mL_tubes",
    nickname: "",
    display_name: "test tuberack 65 mL tubes [2]",
    id: 2,
  },
  {
    name: "reservoir_5",
    labware_name: "test slot 3 mL",
    labware_type: "reservoir",
    labware_id: "test_slot_3_mL",
    nickname: "",
    display_name: "test slot 3 mL [5]",
    id: 5,
  },
];
const sampleCachedProtocolPipettes = [
  {
    name: "(Left) 8 Channel 300 uL - Gen 2",
    category: "left_pipette",
    is_multichannel: true,
  },
  {
    name: "(Right) Single Channel 300 uL - Gen 1",
    category: "right_pipette",
    is_multichannel: true,
  },
];

describe("UnshallowMounted ArrayTransferConfigForm", () => {
  it("has data", () => {
    expect(typeof ArrayTransferConfigForm.data).toBe("function");
  });
});

describe("shallowMounted ArrayTransferConfigForm", () => {
  let getters;
  let store;
  let wrapper;

  beforeEach(() => {
    getters = {
      cachedProtocolSteps: () => sampleCachedProtocolSteps,
      cachedProtocolLiquidClasses: () => sampleCachedProtocolLiquidClasses,
      nonTipRackLabwares: () => sampleNonTipRackLabwares,
      cachedProtocolPipettes: () => sampleCachedProtocolPipettes,
    };

    store = new Vuex.Store({
      getters,
    });

    wrapper = shallowMount(ArrayTransferConfigForm, {
      store,
      localVue,
      propsData: {
        formType: sampleFormType,
        stepID: sampleStepID,
      },
    });
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
