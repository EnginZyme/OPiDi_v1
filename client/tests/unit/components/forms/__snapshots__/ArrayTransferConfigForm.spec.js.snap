// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallowMounted ArrayTransferConfigForm renders correctly 1`] = `
<div
  class="step_config_form"
>
  <p
    class="step_config_form__title"
  >
    Array Transfer
  </p>
   
  <div
    class="step_config_form__body"
  >
    <div
      class="step_config_form__body__content"
    >
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          
          Step Name
        
        </p>
         
        <inputtext
          class="step_config_form__body__content__input--long"
          type="text"
        />
      </div>
       
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          Pipette
        </p>
         
        <dropdown
          class="step_config_form__body__content__dropdown"
          optionlabel="name"
          options="[object Object]"
          placeholder="Select option"
        />
      </div>
       
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          
          Tips Strategy
        
        </p>
         
        <dropdown
          class="step_config_form__body__content__dropdown"
          optionlabel="name"
          options="[object Object],[object Object],[object Object],[object Object]"
          placeholder="Select option"
        />
      </div>
       
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          
          Liquid Class
        
        </p>
         
        <dropdown
          class="step_config_form__body__content__dropdown"
          optionlabel="name"
          options="[object Object]"
          placeholder="Select option"
        />
      </div>
       
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          
          Source Offset
        
        </p>
         
        <div
          class="step_config_form__body__content__field_group"
        >
          <div>
            <p
              class="step_config_form__body__content__label--min"
            >
              Type
            </p>
             
            <dropdown
              class="step_config_form__body__content__dropdown"
              optionlabel="name"
              options="[object Object],[object Object]"
              placeholder="Select option"
            />
          </div>
           
          <div>
            <p
              class="step_config_form__body__content__label--min"
            >
              
              Value (mm)
            
            </p>
             
            <inputnumber
              class="step_config_form__body__content__input--short"
              maxfractiondigits="5"
              minfractiondigits="2"
            />
          </div>
        </div>
      </div>
       
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          
          Destination Offset
        
        </p>
         
        <div
          class="step_config_form__body__content__field_group"
        >
          <div>
            <p
              class="step_config_form__body__content__label--min"
            >
              Type
            </p>
             
            <dropdown
              class="step_config_form__body__content__dropdown"
              optionlabel="name"
              options="[object Object],[object Object]"
              placeholder="Select option"
            />
          </div>
           
          <div>
            <p
              class="step_config_form__body__content__label--min"
            >
              
              Value (mm)
            
            </p>
             
            <inputnumber
              class="step_config_form__body__content__input--short"
              maxfractiondigits="5"
              minfractiondigits="2"
            />
          </div>
        </div>
      </div>
       
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          
          Array Map Filename
        
        </p>
         
        <div
          class="step_config_form__body__content__file_group"
        >
          <inputtext
            class="step_config_form__body__content__input--stretch"
            readonly="readonly"
          />
           
          <button
            class="step_config_form__body__content__file_upload_button"
            for="uploadmyfile"
            label="Browse"
          />
           
          <input
            accept=".csv"
            id="uploadArrayMapFile"
            type="file"
          />
        </div>
      </div>
       
      <a
        class="step_config_form__body__content__file_hyperlink"
        download=""
        href="/assets/array_map_template_simplified.xlsx"
      >
        Download Template
      </a>
       
      <button
        class="step_config_form__body__content__colour_button p-button-raised p-button-success"
        label="Update Array Mapping"
      />
    </div>
  </div>
   
  <arraymapform-stub
    pipette="[object Object]"
    stepid="8"
  />
</div>
`;
